moveit_core/smoothing_plugins/src/butterworth_filter.cpp

```diff
@@ -0,0 +1,102 @@
+#include <moveit/smoothing_plugins/butterworth_filter.h>
+
+namespace smoothing_plugins
+{
+namespace
+{
+constexpr double EPSILON = 1e-9;
+}
+
+ButterworthFilter::ButterworthFilter(double low_pass_filter_coeff)
+  : previous_measurements_{ 0., 0. }
+  , previous_filtered_measurement_(0.)
+  , scale_term_(1. / (1. + low_pass_filter_coeff))
+  , feedback_term_(1. - low_pass_filter_coeff)
+{
+  // guarantee this doesn't change because the logic below depends on this length implicity
+  static_assert(ButterworthFilter::FILTER_LENGTH == 2,
+                "smoothing_plugins::ButterworthFilter::FILTER_LENGTH should be 2");
+
+  if (std::isinf(feedback_term_))
+    throw std::length_error("smoothing_plugins::ButterworthFilter: infinite feedback_term_");
+
+  if (std::isinf(scale_term_))
+    throw std::length_error("smoothing_plugins::ButterworthFilter: infinite scale_term_");
+
+  if (low_pass_filter_coeff < 1)
+    throw std::length_error(
+        "smoothing_plugins::ButterworthFilter: Filter coefficient < 1. makes the lowpass filter unstable");
+
+  if (std::abs(feedback_term_) < EPSILON)
+    throw std::length_error(
+        "smoothing_plugins::ButterworthFilter: Filter coefficient value resulted in feedback term of 0");
+}
+
+double ButterworthFilter::filter(double new_measurement)
+{
+  // Push in the new measurement
+  previous_measurements_[1] = previous_measurements_[0];
+  previous_measurements_[0] = new_measurement;
+
+  double new_filtered_measurement = scale_term_ * (previous_measurements_[1] + previous_measurements_[0] -
+                                                   feedback_term_ * previous_filtered_measurement_);
+
+  // Store the new filtered measurement
+  previous_filtered_measurement_ = new_filtered_measurement;
+
+  return new_filtered_measurement;
+}
+
+void ButterworthFilter::reset(const double data)
+{
+  previous_measurements_[0] = data;
+  previous_measurements_[1] = data;
+  previous_filtered_measurement_ = data;
+}
+
+bool ButterworthFilterPlugin::initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model,
+                                         size_t num_joints)
+{
+  node_ = node;
+  num_joints_ = num_joints;
+
+  for (std::size_t i = 0; i < num_joints_; ++i)
+  {
+    // Low-pass filters for the joint positions
+    // TODO(andyz): read the parameter from yaml
+    position_filters_.emplace_back(1.5);
```

Comment:
It looks like you are now reading this in but not using it here

Reply:
Nevermind ^ after looking at this again. But I think we'll be parsing this from the node not the yaml, correct?

Reply:
Meh, I look at that as one and the same thing since yaml is usually converted to parameters. OK, updated it.